server/game/gameSystems/DeckSearchSystem.ts

```diff
@@ -0,0 +1,284 @@
+
+        const properties = this.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
+        const player = properties.player || context.player;
+        const event = this.generateEvent(player, context, additionalProperties) as DeckSearchEvent;
+        const amount = event.amount > -1 ? event.amount : this.getDeck(player).length;
+        let cards = this.getDeck(player).slice(0, amount);
+        if (event.amount === -1) {
+            cards = cards.filter((card) => properties.cardCondition(card, context));
+        }
+        events.push(event);
+        this.selectCard(event, additionalProperties, cards, new Set());
+
+        return events;
+    }
+
+    private getNumCards(numCards: Derivable<number>, context: AbilityContext): number {
+        return typeof numCards === 'function' ? numCards(context) : numCards;
+    }
+
+    private getAmount(amount: Derivable<number>, context: AbilityContext): number {
+        return typeof amount === 'function' ? amount(context) : amount;
+    }
+
+    private shouldShuffle(shuffle: Derivable<boolean>, context: AbilityContext): boolean {
+        return typeof shuffle === 'function' ? shuffle(context) : shuffle;
+    }
+
+    private getDeck(player: Player): Card[] {
+        return player.drawDeck;
+    }
+
+    private selectCard(event: DeckSearchEvent, additionalProperties: unknown, cards: Card[], selectedCards: Set<Card>): void {
+        const context: AbilityContext = event.context;
+        const properties = this.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
+        const canCancel = properties.targetMode !== TargetMode.Exactly;
+        let selectAmount = 1;
+        const choosingPlayer = properties.choosingPlayer || event.player;
+
+        if (properties.targetMode === TargetMode.UpTo || properties.targetMode === TargetMode.UpToVariable) {
+            selectAmount = this.getNumCards(properties.selectCardCount, context);
+        }
+        if (properties.targetMode === TargetMode.Single) {
+            selectAmount = 1;
+        }
+        if (properties.targetMode === TargetMode.Exactly || properties.targetMode === TargetMode.ExactlyVariable) {
+            selectAmount = this.getNumCards(properties.selectCardCount, context);
+        }
+        if (properties.targetMode === TargetMode.Unlimited) {
+            selectAmount = -1;
+        }
+
+        let title = properties.activePromptTitle;
+        if (!properties.activePromptTitle) {
+            title = 'Select a card' + (properties.reveal ? ' to reveal' : '');
+            if (selectAmount < 0 || selectAmount > 1) {
+                title =
+                    `Select ${selectAmount < 0 ? 'all' : 'up to ' + selectAmount} cards` +
+                    (properties.reveal ? ' to reveal' : '');
+            }
+        }
+
+        if (properties.shuffle) {
+            cards.sort((a, b) => a.name.localeCompare(b.name));
```

Comment:
Two comments:
1. What's the shuffling method here? It looks like it's sorting by name
2. I think we probably need to write the `ShuffleDeckSystem` to handle this unfortunately (ditto for line 214 and `MoveCardSystem` line 61)

Reply:
I'm not sure why it name sorts here, it looks like it does the shuffling separately later on.

How would I call the shuffle system from here?

Reply:
Discussed in chat - see `executeHandler` in `PlayUnitAction` and `InitiateAttackAction` for examples of triggering a system directly from code using an event window.